File: src/node.rs
// Purpose: Contains the Node struct and its implementation
use crate::Jxx;
use crate::SubLattice;

pub struct Node {
    pub index: i32,              // Index of the node
    pub right: i32,              // 2D index of the right node
    pub bottom: i32,             // 2D index of the bottom node
    pub btm_right: i32,          // 2D index of the bottom right node
    pub layer_up: i32,           // 3D index of the layer up node
    pub spin: bool,              // Spin of the node, true: up, false: down
    pub sub_lattice: SubLattice, // Sub-lattice color of the node
    pub j_right: f64,            // J_{i,j} of current node and right node
    pub j_bottom: f64,           // J_{i,j} of current node and bottom node
    pub j_btm_right: f64,        // J_{i,j} of current node and bottom right node
    pub j_layer_up: f64,         // J_{i,j} of current node and layer up node
}

impl Node {
    pub fn new(
        index: i32,
        right: i32,
        bottom: i32,
        btm_right: i32,
        layer_up: i32,
        sub_lattice: SubLattice,
        jxx: &Jxx,
    ) -> Node {
        let j_value = jxx.j as f64; // Default J_{i,j} value
        Node {
            index,
            right,
            bottom,
            btm_right,
            layer_up,
            spin: false,
            sub_lattice,
            j_right: j_value,
            j_bottom: j_value,
            j_btm_right: j_value,
            j_layer_up: jxx.jl as f64,
        }
    }
    pub fn print_info(&self) {
        println!(
            "index: {:>2}; right: {:>2}, {:>5}; bottom: {:>2}, {:>5}; btm_right: {:>2}, {:>5}; layer_up: {:>2}, {:>5}; spin: {:>5}; sub_lattice: {:?};",
            self.index,
            self.right,
            self.j_right,
            self.bottom,
            self.j_bottom,
            self.btm_right,
            self.j_btm_right,
            self.layer_up,
            self.j_layer_up,
            self.spin,
            self.sub_lattice,
        );
    }
}

File: ising_model/Node.cpp
#include "Node.h"
#include <iostream>

// Constructors

Node::Node()
    : index(0), right(0), bottom(0), btm_right(0), layer_up(0), spin(UP), sub_lattice(RED), j_right(1.0), j_bottom(1.0), j_btm_right(1.0),
      j_layer_up(1.0) {
    return;
}

Node::Node(int index, int right, int bottom, int btm_right, int layer_up, SubLattice sub_lattice, Jxx jxx)
    : index(index), right(right), bottom(bottom), btm_right(btm_right), layer_up(layer_up), sub_lattice(sub_lattice) {
    this->spin = UP;
    this->j_right = jxx.j;
    this->j_bottom = jxx.j;
    this->j_btm_right = jxx.j;
    this->j_layer_up = jxx.jl;

    return;
}

Node::Node(int index, int right, int bottom, int btm_right, int layer_up, bool spin, SubLattice sub_lattice, double j_right, double j_bottom,
           double j_btm_right, double j_layer_up)
    : index(index), right(right), bottom(bottom), btm_right(btm_right), layer_up(layer_up), spin(UP), sub_lattice(sub_lattice), j_right(j_right),
      j_bottom(j_bottom), j_btm_right(j_btm_right), j_layer_up(j_layer_up) {
    return;
}

// Getters

int Node::getIndex() const { return index; }

int Node::getRight() const { return right; }

int Node::getBottom() const { return bottom; }

int Node::getBtmRight() const { return btm_right; }

int Node::getLayerUp() const { return layer_up; }

Spin Node::getSpin() const { return spin; }

SubLattice Node::getSubLattice() const { return sub_lattice; }

double Node::getJRight() const { return j_right; }

double Node::getJBottom() const { return j_bottom; }

double Node::getJBtmRight() const { return j_btm_right; }

double Node::getJLayerUp() const { return j_layer_up; }

// Setters

void Node::setIndex(const int index) { this->index = index; }

void Node::setRight(const int right) { this->right = right; }

void Node::setBottom(const int bottom) { this->bottom = bottom; }

void Node::setBtmRight(const int btm_right) { this->btm_right = btm_right; }

void Node::setLayerUp(const int layer_up) { this->layer_up = layer_up; }

void Node::setSpin(const Spin spin) { this->spin = spin; }

void Node::setSubLattice(const SubLattice sub_lattice) { this->sub_lattice = sub_lattice; }

void Node::setJRight(const double j_right) { this->j_right = j_right; }

void Node::setJBottom(const double j_bottom) { this->j_bottom = j_bottom; }

void Node::setJBtmRight(const double j_btm_right) { this->j_btm_right = j_btm_right; }

void Node::setJLayerUp(const double j_layer_up) { this->j_layer_up = j_layer_up; }

// Printers

void Node::printNode() const {
    printf("========== Node %d ==========\n", index);
    printf("Right: %d\tBottom: %d\tBtmRight: %d\tLayerUp: %d\n", right, bottom, btm_right, layer_up);
    printf("Spin: %d\tSubLattice: %d\n", spin, sub_lattice);
    printf("JRight: %f\tJBottom: %f\tJBtmRight: %f\tJLayerUp: %f\n", j_right, j_bottom, j_btm_right, j_layer_up);
    printf("\n");
    return;
}
